### PR TITLE
Add GlitchFilter

### PR DESCRIPTION
**Pull request by finscn**
More details at : https://github.com/pixijs/pixi-filters/issues/100

### CODE REVIEW

**Comment by bigtimebuddy** [comment]
Would recommend converting this: 
`constructor(slices?:number, offset?:number, direction?:number, options?:GlitchOptions);`
to this:
`constructor(options?:GlitchOptions);`

The arguments `slices`, `offset` and `direction` should be in options.

**Comment by finscn** [comment]
offset & direction could move to options , but `slices` can't.
My english is too bad , I can't explain my mind. sorry.
you read the code , you could find the answer.

**Comment by bigtimebuddy** [comment]
@finscn, logic should follow design. We need to start with a better API and that means not using “Object.assign(this, default)”

**Comment by finscn** [comment]
@bigtimebuddy , I see & I've refactored it.

**Comment by bigtimebuddy** [comment]
Much better :+1:

File: filters/glitch/src/GlitchFilter.js [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+ *
+ * @class
+ * @extends PIXI.Filter
+ * @memberof PIXI.filters
+ * @param {object} options - The more optional parameters of the filter.
+ * @param {number} [options.slices=5] - The count of slices.
+ * @param {number} [options.offset=100] - The max-offset of slices.
+ * @param {number} [options.direction=0] - The angle in degree of the offset of slices.
+ * @param {number} [options.fillMode=0] - The fill mode of the space after the offset.
+ *                 0: TRANSPARENT; 1: ORIGINAL; 2: LOOP; 3: CLAMP; 4: MIRROR.
```
Comment (bigtimebuddy) [inline filters/glitch/src/GlitchFilter.js]: Could you add `GlitchFilter.TRANSPARENT`, `GlitchFilter.ORIGINAL` etc as static constants that can be referenced by developers? Easier to use than remember an integer.

File: filters/glitch/src/GlitchFilter.js [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+            seed: 0.5,
+            red: [0, 0],
+            green: [0, 0],
+            blue: [0, 0],
+            minSliceWidth: 8,
+            displacementMapSize: 512,
+            displacementMap: null,
+        };
+
+        Object.assign(_options, options);
```
Comment (bigtimebuddy) [inline filters/glitch/src/GlitchFilter.js]: Better way to do this is to reassign `options`. Avoids `_options` which is confusing.
```js
options = Object.assign({
    slices: 5,
    offset: 100,
    direction: 0,
    fillMode: 0,
    average: false,
    seed: 0.5,
    red: [0, 0],
    green: [0, 0],
    blue: [0, 0],
    minSliceWidth: 8,
    displacementMapSize: 512,
    displacementMap: null,
}, options);
```

File: filters/glitch/src/GlitchFilter.js [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+    setSlicesWidth(slicesWidth) {
+        const len = Math.min(this._slices, slicesWidth.length);
+
+        for (let i = 0; i < len; i++){
+            this.slicesWidth[i] = slicesWidth[i];
+        }
+    }
+
+    setSlicesOffset(slicesOffset) {
```
Comment (bigtimebuddy) [inline filters/glitch/src/GlitchFilter.js]: `setSlicesOffset`, `setSlicesWidth`, `updateDisplacementMap`, `refresh`, `shuffle`, and `initSlicesOffset` need documentation. What of these methods are private or public?

Comment (finscn) [reply filters/glitch/src/GlitchFilter.js]: They are public , and for Advanced Usage.
User could custom the effect of giltch (default effect is random generation).

But I don't know how to explain them....

Comment (bigtimebuddy) [reply filters/glitch/src/GlitchFilter.js]: Could you at least try to explain and add the `/** */` JSDoc blocks? I can help cleanup the language.

File: filters/glitch/src/GlitchFilter.js (hunk 1 above)
Comment (finscn) [reply filters/glitch/src/GlitchFilter.js]: How to define static property ? 
I use 
```
export default class GlitchFilter extends PIXI.Filter {

    static TRANSPARENT = 0;
    static ORIGINAL = 1;
    static LOOP = 2;
    static CLAMP = 3;
    static MIRROR = 4;
```
but Failed

Comment (bigtimebuddy) [reply filters/glitch/src/GlitchFilter.js]: You can define them outside of the class. Or as `static get` methods.

Comment (finscn) [reply filters/glitch/src/GlitchFilter.js]: have to use ```
    GlitchFilter.TRANSPARENT = 0;
    GlitchFilter.ORIGINAL = 1;
    GlitchFilter.LOOP = 2;
    GlitchFilter.CLAMP = 3;
    GlitchFilter.MIRROR = 4;
```
So ugly...

Comment (finscn) [reply filters/glitch/src/GlitchFilter.js]: DONE.

File: filters/glitch/src/GlitchFilter.js (hunk 2 above)
Comment (finscn) [reply filters/glitch/src/GlitchFilter.js]: DONE

**Comment by bigtimebuddy** [comment]
@finscn Could you review my edits? Basically:

I renamed and simplified a bunch of variables and methods. 

I added more documentation, but I still don't know how to describe `average` or what it does.

I removed the ability to use your own displacement map texture. This functionality was half-baked when I tested it, I would rather add it back when there's a clear use-case. Otherwise, it complicates too much.

File: filters/glitch/src/GlitchFilter.js [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+     *
+     * @private
+     */
+    _randomizeSizes() {
+        const arr = this._sizes;
+        const last = this._slices - 1;
+        const size = this.sampleSize;
+        const min = Math.min(this.minSize / size, 0.9 / this._slices);
+
+        if (this.average) {
```
Comment (bigtimebuddy) [inline filters/glitch/src/GlitchFilter.js]: @finscn, I don't understand the benefit of using `average` just different ways to slice up the total size. I'd recommend just picking one. The end result looks fine either way.

Comment (finscn) [reply filters/glitch/src/GlitchFilter.js]: average  will make the widths of slices is almost in same width.
Just almost.  e.g.
if average == false ,  the width maybe  from  1% to 70% .
if average == true ,  the width maybe from  20% to 30%.

File: filters/glitch/src/GlitchFilter.js [hunk 5]
```diff
@@ -33,37 +38,91 @@ export default class GlitchFilter extends PIXI.Filter {
...
+        this.sampleSize = options.sampleSize;
+
+        /**
+         * Internally generated canvas.
+         *
+         * @member {HTMLCanvasElement} _canvas
+         * @private
+         */
+        this._canvas = document.createElement('canvas');
+        this._canvas.width = 8;
```
Comment (finscn) [inline filters/glitch/src/GlitchFilter.js]: I think    4 is  enough & safe.  (2 is not safe)
in the shader , will read the middle of width .

**Comment by finscn** [comment]
I reviewed.  There are 2 request : 
shuffle  &  _redrawTexture  should be public. 

At least , I use them in my game.
The use case  : 
> I don't want to use `random slice width & slice base-offset` , I custom them.
> in game tick , I  shuffle ( then _redrawTexture)  slices by some intervals.
>

**Comment by bigtimebuddy** [comment]
@finscn how about now?

**Comment by finscn** [comment]
Only one suggestion : 
When `_redrawTexture` become public , I hope the name could be changed.  
`Texture` is too abstract.  User maybe can't understand : 
```What's texture? 
where's the texture? 
Why is there a texture in the filter? 
what's it doing?
```
I like  `redrawDisplacementMap` , though , it is long ,  it's easy to understand.

Whatever , your english is better than me. and you know  the habit of English speaking men more clearly then me.
So you decide this thing.

**Comment by bigtimebuddy** [comment]
@finscn, What about just `redraw` instead of `redrawTexture`? I feel like knowing that this filter uses a displacement map approach is not important for understanding how to use it.

**Comment by finscn** [comment]
That's OK.  
And this method is for advanced users.
The advanced users could understand it , whatever how to name it.

**Pull request merged**